File: frontend/src/lib/cube/Move.ts
/**
 * Face enum representing the six faces of a Rubik's Cube
 * Using Singmaster notation
 */
export type Face = 'U' | 'D' | 'L' | 'R' | 'F' | 'B';

/**
 * Direction enum for cube rotations
 * 1: Clockwise 90°
 * -1: Counter-clockwise 90°
 * 2: 180°
 */
export type Direction = 1 | -1 | 2;

/**
 * Represents a single move on the Rubik's Cube
 */
export interface Move {
  face: Face;
  direction: Direction;
}

/**
 * Convert a Move object to Singmaster notation string
 * @param move The move to convert
 * @returns String representation (e.g., "U", "R'", "F2")
 */
export function moveToString(move: Move): string {
  if (move.direction === 1) return move.face;
  if (move.direction === -1) return `${move.face}'`;
  return `${move.face}2`;
}

/**
 * Parse Singmaster notation string to Move object
 * @param s Singmaster notation (e.g., "U", "R'", "F2")
 * @returns Move object
 * @throws Error if string is invalid
 */
export function stringToMove(s: string): Move {
  const face = s[0] as Face;
  
  if (!['U', 'D', 'L', 'R', 'F', 'B'].includes(face)) {
    throw new Error(`Invalid face: ${face}`);
  }
  
  if (s.length === 1) return { face, direction: 1 };
  if (s[1] === "'") return { face, direction: -1 };
  if (s[1] === '2') return { face, direction: 2 };
  
  throw new Error(`Invalid move string: ${s}`);
}

/**
 * Get the inverse of a move
 * @param move The move to inverse
 * @returns The inverse move
 */
export function inverseMove(move: Move): Move {
  if (move.direction === 2) return move; // 180° is self-inverse
  return { ...move, direction: (-move.direction) as Direction };
}

/**
 * Convert array of moves to Singmaster notation string
 * @param moves Array of moves
 * @returns Space-separated string (e.g., "U R' F2 D")
 */
export function movesToString(moves: Move[]): string {
  return moves.map(moveToString).join(' ');
}

/**
 * Parse Singmaster notation string to array of moves
 * @param s Space-separated moves (e.g., "U R' F2 D")
 * @returns Array of Move objects
 */
export function stringToMoves(s: string): Move[] {
  return s.split(' ').filter(Boolean).map(stringToMove);
}

/**
 * Check if a move is valid
 * @param move The move to validate
 * @returns true if move is valid
 */
export function isValidMove(move: any): boolean {
  if (!move || typeof move !== 'object') return false;
  if (!['U', 'D', 'L', 'R', 'F', 'B'].includes(move.face)) return false;
  if (![1, -1, 2].includes(move.direction)) return false;
  return true;
}

File: backend/src/models/move.py
"""
Move model for Rubik's Cube
"""
from dataclasses import dataclass
from enum import Enum
from typing import Literal


class Face(str, Enum):
    """Face enum using Singmaster notation"""
    U = "U"
    D = "D"
    L = "L"
    R = "R"
    F = "F"
    B = "B"


Direction = Literal[1, -1, 2]


@dataclass(frozen=True)
class Move:
    """
    Represents a single move on the Rubik's Cube

    Attributes:
        face: The face to rotate (U/D/L/R/F/B)
        direction: Rotation direction (1=CW 90°, -1=CCW 90°, 2=180°)
    """
    face: Face
    direction: Direction

    def __post_init__(self):
        if self.direction not in (1, -1, 2):
            raise ValueError(f"Direction must be 1, -1, or 2, got {self.direction}")

    def to_string(self) -> str:
        """Convert to Singmaster notation"""
        if self.direction == 1:
            return self.face.value
        elif self.direction == -1:
            return f"{self.face.value}'"
        return f"{self.face.value}2"

    def to_dict(self) -> dict:
        """Convert to dictionary for API responses"""
        return {
            "face": self.face.value,
            "direction": self.direction
        }

    @classmethod
    def from_string(cls, s: str) -> "Move":
        """Parse Singmaster notation to Move"""
        if not s or len(s) < 1:
            raise ValueError(f"Invalid move string: {s}")

        face_char = s[0].upper()
        try:
            face = Face(face_char)
        except ValueError:
            raise ValueError(f"Invalid face: {face_char}")

        if len(s) == 1:
            return cls(face=face, direction=1)
        elif s[1] == "'":
            return cls(face=face, direction=-1)
        elif s[1] == "2":
            return cls(face=face, direction=2)
        else:
            raise ValueError(f"Invalid move string: {s}")

    def inverse(self) -> "Move":
        """Return the inverse move"""
        if self.direction == 2:
            return self  # 180° is self-inverse
        return Move(face=self.face, direction=-self.direction)  # type: ignore


def moves_to_string(moves: list[Move]) -> str:
    """Convert list of moves to space-separated string"""
    return " ".join(move.to_string() for move in moves)


def string_to_moves(s: str) -> list[Move]:
    """Parse space-separated moves string"""
    if not s.strip():
        return []
    return [Move.from_string(part) for part in s.split() if part.strip()]
